Add support for importing a trivial global C++ function

bricknerb opened the pull request:
ASTUnit is owned by `CompileSubcommand`, passed through `Unit` to be populated in `ImportCppFiles()` and used via `SemIR::File`.
When generating the AST, pass `-x c++` args to compile C++ (temporary until we pass args properly).
`Cpp` namespace is marked as a special namespace and has dedicated logic in `LookupNameInExactScope()`.
The logic for importing declarations from C++ to Carbon is in `import_cpp.cpp`, but we're likely to want to refactor this signfiicantly over time as it grows (perhaps a dedicated directory?).

Part of #4666.

bricknerb committed:
Add support for import a trivial global C++ function
ASTUnit is owned by `CompileSubcommand`, passed through `Unit` to be populated in `ImportCppFiles()` and used via `SemIR::File`.
When generating the AST, pass `-x c++` args to compile C++ (temporary until we pass args properly).
`Cpp` namespace is marked as a special namespace and has dedicated logic in `LookupNameInExactScope()`.
The logic for importing declarations from C++ to Carbon is in `import_cpp.cpp`, but we're likely to want to refactor this signfiicantly over time as it grows (perhaps a dedicated directory?).

jonmeow commented on toolchain/sem_ir/file.h:
```diff
@@ -186,6 +187,8 @@ class File : public Printable<File> {
     return import_cpps_;
   }
+  auto cpp_ast() -> clang::ASTUnit* { return cpp_ast_; }
+  auto set_cpp_ast(clang::ASTUnit* cpp_ast) -> void { cpp_ast_ = cpp_ast; }
```
Storing a pointer to the AST may be helpful, but can you add a TODO to move this to the constructor and remove the mutator function? You may be calling `ImportCppFiles` after the `File` is constructed right now, but it should probably be constructed earlier and instead have the AST passed in to the `File` constructor.

bricknerb replied:
Done.

jonmeow commented on toolchain/check/name_lookup.cpp:
```diff
@@ -150,6 +152,18 @@ auto LookupNameInExactScope(Context& context, SemIR::LocId loc_id,
+        ImportNameFromCpp(context, loc_id, scope_id, name_id);
+    if (imported_inst_id.has_value()) {
+      SemIR::ScopeLookupResult result = SemIR::ScopeLookupResult::MakeFound(
+          imported_inst_id, SemIR::AccessKind::Public);
```
Should this have a TODO to enforce access modifiers? You might be working on namespaces right now, but this is generally for name scopes. i.e., similar to the TODO on 149, maybe this will encounter protected/private in type scopes.

zygoloid replied:
The destructor of the `clang::LookupResult` object will issue diagnostics if the named entity isn't accessible. Now you mention it, it'd be good to have a test of that. It will also diagnose the case where the named entity is ambiguous, which'd be nice to test too.

jonmeow replied:
That might work in some cases, but here's what I have in mind:

C++

```c++
class Foo {
 protected:
  auto F() -> void;
};
```

Carbon

```
class Bar {
  extend base: Cpp.Foo;

  // Not an access error.
  fn G[addr self: Self*]() { self->F(); }
}

fn F() {
  var b: Bar;
  // Access error.
  b.F();
}
```

zygoloid replied:
I see. Yeah, I agree, we'll need to map C++ access into Carbon access here.

bricknerb replied:
> The destructor of the `clang::LookupResult` object will issue diagnostics if the named entity isn't accessible. Now you mention it, it'd be good to have a test of that. It will also diagnose the case where the named entity is ambiguous, which'd be nice to test too.

Seems like `clang::LookupResult`'s destructor doesn't diagnose if for `RedeclarationKind::NotForRedeclaration`, which is the default and what I use.
It also seems that access check is only relevant when doing class lookup, which we don't support anyways.
I've added a TODO to support class lookup and to add a check for access (though I don't think I can test this TODO yet).

> I see. Yeah, I agree, we'll need to map C++ access into Carbon access here.

IIUC, we'll do that as part of adding support for importing classes.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -131,10 +139,146 @@ auto ImportCppFiles(Context& context, llvm::StringRef importing_file_path,
   }
 }
 
+static auto ClangLookup(Context& context, SemIR::NameId name_id)
```
Please add function comments to these statics (including ImportFunctionDecl and so on below).

bricknerb replied:
Done.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -113,14 +116,19 @@ static auto AddNamespace(Context& context, PackageNameId cpp_package_id,
 }
 
 auto ImportCppFiles(Context& context, llvm::StringRef importing_file_path,
+                    std::unique_ptr<clang::ASTUnit>* ast,
```
I think we'd pass-by-reference, not by pointer, since the `ast` pointer itself is not stored. That'd also eliminate the CHECK on 127.
```suggestion
                    std::unique_ptr<clang::ASTUnit>& ast,
```

Though, had you considered having `ImportCppFiles` return `std::unique_ptr<clang::ASTUnit>` instead of taking the return value as a parameter?

Also, other style note -- since this is an output parameter, it should typically be [the last parameter](https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs). Though that's moot if you just return the pointer.

bricknerb replied:
Done.

jonmeow commented on toolchain/sem_ir/file.h:
```diff
@@ -301,6 +304,9 @@ class File : public Printable<File> {
   // List of Cpp imports.
   ValueStore<ImportCppId> import_cpps_;
 
+  // The Clang AST to used when looking up `Cpp` names.
```
```suggestion
  // The Clang AST to use when looking up `Cpp` names. Null if there are no `Cpp` imports.
```

bricknerb replied:
Done.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -131,10 +139,146 @@ auto ImportCppFiles(Context& context, llvm::StringRef importing_file_path,
+    -> std::optional<clang::LookupResult> {
+  clang::ASTUnit* ast = context.sem_ir().cpp_ast();
+  CARBON_CHECK(ast);
+  CARBON_CHECK(ast->hasSema());
+  clang::Sema& sema = ast->getSema();
```
```suggestion
  clang::Sema& sema = ast->getSema();
```

Note the implementation of getSema:

```
  bool hasSema() const { return (bool)TheSema; }

  Sema &getSema() const {
    assert(TheSema && "ASTUnit does not have a Sema object!");
    return *TheSema;
  }
```

bricknerb replied:
Done.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -131,10 +139,146 @@ auto ImportCppFiles(Context& context, llvm::StringRef importing_file_path,
+  clang::Sema& sema = ast->getSema();
+
+  auto name = context.names().GetAsStringIfIdentifier(name_id);
+  CARBON_CHECK(name);
```
Please test what happens with lookup for `Cpp.base` -- that should probably diagnose, and could be a TODO, but must not crash.

bricknerb replied:
Added a test and now using `GetIRBaseName()`.

jonmeow replied:
(responding as part of zygoloid's thread on this)

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -131,10 +139,146 @@ auto ImportCppFiles(Context& context, llvm::StringRef importing_file_path,
+    -> SemIR::InstId {
+  if (clang_decl->isVariadic()) {
+    context.TODO(loc_id, "Unsupported: Variadic function");
+    return SemIR::InstId::None;
```
Maybe return `ErrorInst::SingletonInstId` in these instead of `None` -- name lookup seems to have succeeded, it was just an error.

bricknerb replied:
Done.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -131,10 +139,146 @@ auto ImportCppFiles(Context& context, llvm::StringRef importing_file_path,
+          .param_patterns_id = SemIR::InstBlockId::Empty,
+          .call_params_id = SemIR::InstBlockId::Empty,
+          .is_extern = false,
+          .extern_library_id = SemIR::LibraryNameId ::None,
```
```suggestion
          .extern_library_id = SemIR::LibraryNameId::None,
```

I'm surprised clang-format didn't fix this...?

bricknerb replied:
Done.
Me too.
I've noticed such issues when auto-formatting the code, but I guess I missed this one.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -131,10 +139,146 @@ auto ImportCppFiles(Context& context, llvm::StringRef importing_file_path,
+      context, SemIR::LocIdAndInst(Parse::NodeId::None, function_decl));
+
+  auto function_info = SemIR::Function{
+      SemIR::EntityWithParamsBase{
```
```suggestion
     {
```

Here and on line 225, please remove the type. See "initialization" notes at https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/cpp_style_guide.md#syntax-and-formatting

bricknerb replied:
Done.

zygoloid commented on toolchain/check/import_cpp.cpp:
```diff
@@ -131,10 +139,146 @@ auto ImportCppFiles(Context& context, llvm::StringRef importing_file_path,
+  clang::Sema& sema = ast->getSema();
+
+  auto name = context.names().GetAsStringIfIdentifier(name_id);
+  CARBON_CHECK(name);
```
What happens if we try to look up a non-identifier Carbon name, such as via `Cpp.Self` or `Cpp.base`? Perhaps we should produce a lookup failure for non-identifier names for now.

bricknerb replied:
Added a test and now using `GetIRBaseName()`.

jonmeow replied:
My general understanding is we want to require raw keyword syntax in this sort of situation: i.e., `Cpp.r#base`, with `Cpp.base` being an error. So `GetAsStringIfIdentifier ` was correct, rather than `GetIRBaseName` (which I think mainly exists for debug output).

bricknerb replied:
Done.

zygoloid reviewed (approved):
Very exciting, this looks like a great step :)

jonmeow reviewed (commented):
Similarly, nice to see progress here! I realize I missed the publish button though.

bricknerb committed:
Fix formatting

bricknerb committed:
Style: Avoid typename in braced initialization

bricknerb committed:
When finding a name in `Cpp`, but the found declaration is not supported, return error `InstId` to avoid diagnosing on name not found

bricknerb committed:
Add a TODO to remove `File::set_cpp_ast()`

bricknerb committed:
TODO for `isClassLookup()`

bricknerb committed:
Add function comments

bricknerb committed:
Initialize `Unit.cpp_ast` to null by default.
It's not initialized everywhere and no warning is given.
https://github.com/carbon-language/carbon-lang/blob/43b9969058b3de2e0b05896ef8e4489c1dc85cb6/toolchain/language_server/context.cpp#L146

bricknerb commented on toolchain/sem_ir/file.h:
```diff
@@ -301,6 +304,9 @@ class File : public Printable<File> {
   ValueStore<ImportCppId> import_cpps_;
 
+  // The Clang AST to used when looking up `Cpp` names.
+  clang::ASTUnit* cpp_ast_ = nullptr;
```
Added this initialization since it's not initialized everywhere and no warning is given.
https://github.com/carbon-language/carbon-lang/blob/43b9969058b3de2e0b05896ef8e4489c1dc85cb6/toolchain/language_server/context.cpp#L146

bricknerb committed:
Change `ImportCppFiles()` to return the AST instead of using a return value as a parameter

bricknerb committed:
Improve `File::cpp_ast_` comment

bricknerb committed:
Remove unnecessary `CARBON_CHECK`

bricknerb committed:
Handle special names by using `GetIRBaseName()`

bricknerb committed:
Add a test for using a the wrong name when calling a `Cpp` function

bricknerb committed:
Update test name to include `fail_`

bricknerb commented on toolchain/check/testdata/interop/cpp/no_prelude/function_decl.carbon:
```diff
@@ -18,18 +18,462 @@ library "[[@TEST_NAME]]";
+  Cpp.foo();
+}
+
+// --- fail_import_function_decl_use_different_name.carbon
```
Added this test because it seems to be missing.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -61,7 +62,9 @@ static auto GenerateAst(Context& context, llvm::StringRef importing_file_path,
-      GenerateCppIncludesHeaderCode(context, imports), {},
+      GenerateCppIncludesHeaderCode(context, imports),
+      {// Parse C++ (and not C)
+       "-x", "c++"},
```
```suggestion
      // Parse C++ (and not C).
      {"-x", "c++"},
```

Per style: "Only use line comments (with //, not /* ... */), on a line by themselves"
https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/cpp_style_guide.md

bricknerb replied:
Done.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -131,10 +136,157 @@ auto ImportCppFiles(Context& context, llvm::StringRef importing_file_path,
+  return std::move(generated_ast);
+}
+
+// Lookups the given name in the Clang AST. Returns the lookup result if lookup
```
```suggestion
// Looks up the given name in the Clang AST. Returns the lookup result if lookup
```

Small grammar thing.

bricknerb replied:
Done.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -141,64 +138,75 @@ auto ImportCppFiles(Context& context, llvm::StringRef importing_file_path,
+      lookup, ast->getASTContext().getTranslationUnitDecl());
+
+  if (lookup.isClassLookup()) {
+    // TODO: When support class lookup, also check access.
```
```suggestion
    // TODO: To support class lookup, also return the AccessKind for storage.
```

To be sure, access can't be checked here; that could lead to something like a `protected` member being incorrectly hidden, when the caller doesn't have access. Instead the AccessKind needs to be part of the returned information so that it can be stored in the lookup table (instead of `AccessKind::Public`), and then used by lookup to determine access.

(also, small grammar thing -- "to support" or "when supporting")

bricknerb replied:
Done.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -141,64 +138,75 @@ auto ImportCppFiles(Context& context, llvm::StringRef importing_file_path,
+
+  if (lookup.isClassLookup()) {
+    // TODO: When support class lookup, also check access.
+    context.TODO(loc_id, "Unsupported: Lookup in Class");
```
Maybe (given the "unsupported" string):

```suggestion
    context.TODO(loc_id, "Unsupported: Lookup in Class");
    return std::nullopt;
```

bricknerb replied:
Done.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -131,10 +136,157 @@ auto ImportCppFiles(Context& context, llvm::StringRef importing_file_path,
+  CARBON_CHECK(ast);
+  clang::Sema& sema = ast->getSema();
+
+  llvm::StringRef name = context.names().GetIRBaseName(name_id);
```
```suggestion
  std::optional<llvm::StringRef> name = context.names().GetAsStringIfIdentifier(name_id);
  if (!name) {
    // Special names never exist in C++ code.
    return std::nullopt;
  }
```

(see comment)

bricknerb replied:
Done.
I thought of using a dedicated error for this use case to point out this is a special name (and maybe even tell the user how to fix), but keeping as is.
See https://github.com/carbon-language/carbon-lang/pull/5033/commits/16867c9578f77e5eba2fd8c2ff2765e584e53b75

jonmeow reviewed (approved):
If you want to address final comments and merge, that seems fine. Note I've suggested a significant change not to use `GetIRBaseName` (explanation: https://github.com/carbon-language/carbon-lang/pull/5033#discussion_r1974456129)

bricknerb committed:
Improve class lookup TODO

bricknerb committed:
Do not allow using special names in `Cpp`. Use raw keyword syntax instead.

bricknerb committed:
Fix style and grammar.

bricknerb committed:
Treat special names inside `Cpp` as not existing

bricknerb commented:
> If you want to address final comments and merge, that seems fine. Note I've suggested a significant change not to use `GetIRBaseName` (explanation: [#5033 (comment)](https://github.com/carbon-language/carbon-lang/pull/5033#discussion_r1974456129))

Thanks!
I've addressed the final comments and will try to merge.
